src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,174 @@
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        boolean::multiply_all_shares,
+        context::Context,
+        sort::MultiBitPermutationStep::{MultiplyAcrossBits, Sop},
+        RecordId,
+    },
+    secret_sharing::SecretSharing,
+};
+
+use futures::future::try_join_all;
+
+/// This is an implementation of `GenMultiBitSort` (Algorithm 11) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+///
+/// Protocol to compute a secret sharing of a permutation, after sorting on multiple bits `num_multi_bits`.
+/// At a high level, the protocol works as follows:
+/// 1. Start with a vector of list of `L*n` secret shares `[[x1_1 ... x1_n], .. , [xL_1 ... xL_n]]` where each is a secret sharing of either zero or one.
+/// 2. For j in 0 to 2 pow `num_multi_bits`
+///    i. Get binary representation of j (Bk)
+///    ii. For i in `num_multi_bits`
+///      a. Locally compute `mult_inputs` as (Bk * k + (1-Bk)(1- k))
+///   iii. Multiply all `mult_inputs` for this j
+/// 4. For j in 0 to 2 pow `num_multi_bits`
+///    i. For each record
+///       a. Calculate accumulated `total_sum` = s + `mult_output`
+/// 5. Compute the final output using sum of products executed in parallel for each record.
+#[allow(dead_code)]
+pub async fn multi_bit_permutation_single<
+    'a,
+    F: Field,
+    S: SecretSharing<F>,
+    C: Context<F, Share = S>,
+>(
+    ctx: C,
+    input: &[Vec<S>],
+) -> Result<Vec<S>, Error> {
+    let share_of_one = ctx.share_of_one();
+    let num_multi_bits = input.len();
+    let num_recs = input[0].len();
+    let twonminusone = 2 << (num_multi_bits - 1);
+
+    let mut mult_outputs = Vec::with_capacity(twonminusone);
+    for j in 0..twonminusone {
```

Comment:
All this can definitely be written in a more compact way. Need some help to make it more concise.